File: frontend/src/features/customer/mapper/customerSiteBillingRateMapper.ts
import type {
  CustomerSiteBillingRateRequest,
  CustomerSiteBillingRateResponse,
} from '../types/customerApiTypes'

import type {
  CustomerSiteBillingRate,
} from '../types/customerTypes'

function toStringValue(
  value: string | null | undefined,
): string {
  return value ?? ''
}

function toNullableString(
  value: string | null | undefined,
): string | null {
  const normalized = value?.trim()

  return normalized
    ? normalized
    : null
}

function toNumberValue(
  value: number | null | undefined,
): number {
  return Number(value ?? 0)
}

function toNullableId(
  value: number | null | undefined,
): number | null {
  if (
    value == null
    || value <= 0
  ) {
    return null
  }

  return value
}

export function toCustomerSiteBillingRate(
  response: CustomerSiteBillingRateResponse,
): CustomerSiteBillingRate {
  return {
    id: response.id,
    customerSiteId:
      response.customerSiteId,

    deleteSelected: false,

    jobCode:
      toStringValue(response.jobCode),

    jobName:
      toStringValue(response.jobName),

    siteRoleCode:
      toStringValue(
        response.siteRoleCode,
      ) || 'GENERAL',

    siteRoleName:
      toStringValue(
        response.siteRoleName,
      ) || '一般',

    billingUnit:
      response.billingUnit ?? 'DAILY',

    baseUnitPrice:
      toNumberValue(
        response.baseUnitPrice,
      ),

    overtimeUnitPrice:
      toNumberValue(
        response.overtimeUnitPrice,
      ),

    nightUnitPrice:
      toNumberValue(
        response.nightUnitPrice,
      ),

    holidayUnitPrice:
      toNumberValue(
        response.holidayUnitPrice,
      ),

    commuteUnitPrice:
      toNumberValue(
        response.commuteUnitPrice,
      ),

    effectiveFrom:
      toStringValue(
        response.effectiveFrom,
      ),

    effectiveTo:
      response.effectiveTo ?? null,

    displayOrder:
      response.displayOrder ?? 1,

    activeFlag:
      response.activeFlag ?? true,

    note:
      toStringValue(response.note),

    _isNew: false,
    _isUpdated: false,
    _isDeleted: false,
  }
}

export function toCustomerSiteBillingRateRequest(
  row: CustomerSiteBillingRate,
): CustomerSiteBillingRateRequest {
  const customerSiteId =
    toNullableId(
      row.customerSiteId,
    )

  if (customerSiteId == null) {
    throw new Error(
      '現場は必須です。',
    )
  }

  return {
    id:
      toNullableId(row.id),

    customerSiteId,

    jobCode:
      row.jobCode.trim(),

    jobName:
      row.jobName.trim(),

    siteRoleCode:
      row.siteRoleCode.trim()
      || 'GENERAL',

    siteRoleName:
      row.siteRoleName.trim()
      || '一般',

    billingUnit:
      row.billingUnit ?? 'DAILY',

    baseUnitPrice:
      toNumberValue(
        row.baseUnitPrice,
      ),

    overtimeUnitPrice:
      toNumberValue(
        row.overtimeUnitPrice,
      ),

    nightUnitPrice:
      toNumberValue(
        row.nightUnitPrice,
      ),

    holidayUnitPrice:
      toNumberValue(
        row.holidayUnitPrice,
      ),

    commuteUnitPrice:
      toNumberValue(
        row.commuteUnitPrice,
      ),

    effectiveFrom:
      row.effectiveFrom,

    effectiveTo:
      toNullableString(
        row.effectiveTo,
      ),

    displayOrder:
      row.displayOrder ?? 1,

    activeFlag:
      row.activeFlag ?? true,

    note:
      toNullableString(row.note),

    _isNew:
      row._isNew,

    _isUpdated:
      row._isUpdated,

    _isDeleted:
      row._isDeleted,
  }
}

export function createEmptyCustomerSiteBillingRate(
  customerSiteId: number | null,
  displayOrder = 1,
): CustomerSiteBillingRate {
  return {
    id: -Date.now(),

    customerSiteId:
      customerSiteId ?? -1,

    deleteSelected: false,

    jobCode: '',
    jobName: '',

    siteRoleCode: 'GENERAL',
    siteRoleName: '一般',

    billingUnit: 'DAILY',

    baseUnitPrice: 0,
    overtimeUnitPrice: 0,
    nightUnitPrice: 0,
    holidayUnitPrice: 0,
    commuteUnitPrice: 0,

    effectiveFrom:
      new Date()
        .toISOString()
        .slice(0, 10),

    effectiveTo: null,

    displayOrder,
    activeFlag: true,

    note: '',

    _isNew: true,
    _isUpdated: false,
    _isDeleted: false,
  }
}

File: backend/src/main/java/com/project/backend/features/customer/mapper/CustomerSiteBillingRateMapper.java
package com.project.backend.features.customer.mapper;

import java.math.BigDecimal;

import org.springframework.stereotype.Component;

import com.project.backend.features.customer.dto.CustomerSiteBillingRateRequest;
import com.project.backend.features.customer.dto.CustomerSiteBillingRateResponse;
import com.project.backend.features.customer.entity.CustomerSite;
import com.project.backend.features.customer.entity.CustomerSiteBillingRate;
import com.project.backend.features.customer.enums.CustomerBillingUnit;

@Component
public class CustomerSiteBillingRateMapper {

    public CustomerSiteBillingRate toEntity(
            CustomerSite customerSite,
            CustomerSiteBillingRateRequest request
    ) {
        CustomerSiteBillingRate entity =
                new CustomerSiteBillingRate();

        entity.setCustomerSite(customerSite);

        apply(
                entity,
                request
        );

        return entity;
    }

    public void apply(
            CustomerSiteBillingRate entity,
            CustomerSiteBillingRateRequest request
    ) {
        if (request == null) {
            throw new IllegalArgumentException(
                    "現場請求単価リクエストは必須です。"
            );
        }

        entity.setJobCode(
                trimToNull(request.jobCode())
        );

        entity.setJobName(
                trimToNull(request.jobName())
        );

        entity.setSiteRoleCode(
                trimToNull(request.siteRoleCode()) == null
                        ? "GENERAL"
                        : request.siteRoleCode().trim()
        );

        entity.setSiteRoleName(
                trimToNull(request.siteRoleName()) == null
                        ? "一般"
                        : request.siteRoleName().trim()
        );

        entity.setBillingUnit(
                request.billingUnit() == null
                        ? CustomerBillingUnit.DAILY
                        : request.billingUnit()
        );

        entity.setBaseUnitPrice(
                toZero(request.baseUnitPrice())
        );

        entity.setOvertimeUnitPrice(
                toZero(request.overtimeUnitPrice())
        );

        entity.setNightUnitPrice(
                toZero(request.nightUnitPrice())
        );

        entity.setHolidayUnitPrice(
                toZero(request.holidayUnitPrice())
        );

        entity.setCommuteUnitPrice(
                toZero(request.commuteUnitPrice())
        );

        entity.setEffectiveFrom(
                request.effectiveFrom()
        );

        entity.setEffectiveTo(
                request.effectiveTo()
        );

        entity.setDisplayOrder(
                request.displayOrder() == null
                        ? 1
                        : request.displayOrder()
        );

        entity.setActiveFlag(
                request.activeFlag() == null
                        ? true
                        : request.activeFlag()
        );

        entity.setNote(
                trimToNull(request.note())
        );
    }

    public CustomerSiteBillingRateResponse toResponse(
            CustomerSiteBillingRate entity
    ) {
        return new CustomerSiteBillingRateResponse(
                entity.getId(),
                entity.getCustomerSite().getId(),

                entity.getJobCode(),
                entity.getJobName(),

                entity.getSiteRoleCode(),
                entity.getSiteRoleName(),

                entity.getBillingUnit(),

                entity.getBaseUnitPrice(),
                entity.getOvertimeUnitPrice(),
                entity.getNightUnitPrice(),
                entity.getHolidayUnitPrice(),
                entity.getCommuteUnitPrice(),

                entity.getEffectiveFrom(),
                entity.getEffectiveTo(),

                entity.getDisplayOrder(),
                entity.getActiveFlag(),
                entity.getNote()
        );
    }

    private BigDecimal toZero(
            BigDecimal value
    ) {
        return value == null
                ? BigDecimal.ZERO
                : value;
    }

    private String trimToNull(
            String value
    ) {
        if (value == null || value.isBlank()) {
            return null;
        }

        return value.trim();
    }
}